### PR TITLE
[REF] New reporting implementation

### PR DESCRIPTION
**Pull request by Alroniks**
Work is done under task [CGF-23](https://oat-sa.atlassian.net/browse/CGF-23)

Current implementation covers all types of messages and provides methods for checking and getting these messages.

Old classes still used for BC, but marked as deprecated.

### CODE REVIEW

**Review by AndreyNegrov — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] The feature is working correctly on my local machine (if applicable)
- [x] Pull request's target is not `master`

File: common/oatbox/reporting/Report.php [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+            return $this->contains(strtolower(str_replace('contains', '', $name)));
+        }
+
+        return null;
+    }
+
+    /**
+     * Get message of the report
+     *
+     * @return string
```
Comment (boajer) [inline common/oatbox/reporting/Report.php:168]: What additional value `@return` PHPDoc adds on top of strictly defined return type in method signature?

Comment (Alroniks) [reply common/oatbox/reporting/Report.php:168]: Actually, nothing, seems it left after moving code to the new file.

Comment (boajer) [reply common/oatbox/reporting/Report.php:168]: WDYT regarding cleanup such leftovers, in the code you've touched ?

File: test/unit/ReportTest.php [hunk 2]
```diff
@@ -16,142 +15,178 @@
...
+    public function testBasicReport(): void
     {
-        $report = new common_report_Report(common_report_Report::TYPE_SUCCESS, 'test message');
-        $this->assertFalse($report->hasChildren());
-        $this->assertEquals('test message', (string)$report);
-        $this->assertEquals('test message', $report->getMessage());
-        $this->assertEquals(common_report_Report::TYPE_SUCCESS, $report->getType());
+        $report = new Report(Report::TYPE_SUCCESS, 'test message');
+
+        self::assertFalse($report->hasChildren());
```
Comment (hutnikau) [inline test/unit/ReportTest.php:31]: why use `self` instead of `$this`?

**Review by hutnikau — CHANGES_REQUESTED** [review]

File: common/oatbox/reporting/RecursiveReportIterator.php [hunk 3]
```diff
@@ -14,22 +14,36 @@
...
+
 /**
- * An element of a report representing an error that occured
+ * Custom RecursiveIterator for reports.
  *
- * @access public
- * @author Joel Bout, <joel@taotesting.com>
- * @package generis
- * @deprecated
+ * @author Gyula Szucs, <gyula@taotesting.com>
```
[the remark this answers is not in the document]
Comment (Alroniks) [reply common/oatbox/reporting/RecursiveReportIterator.php:30]: Yes, github shows diff a bit weird. ErrorElement was deleted as unused, but RecursiveReportIterator just was moved to a new namespace. The author is the same.

File: test/unit/ReportTest.php (hunk 2 above)
Comment (Alroniks) [reply test/unit/ReportTest.php:31]: PHPUnit provides such methods as static, phpstorm in inspections says that more affective call it as static. I tend to agree with it.

**Comment by hutnikau** [comment]
**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] The feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by hutnikau — APPROVED** [review]

**Pull request merged**